pilot/pkg/dns/dns.go

```diff
@@ -224,40 +235,64 @@ func separateIPtypes(ips []string) (ipv4, ipv6 []net.IP) {
 }
 
 func generateAltHosts(hostname string, nameinfo *nds.NameTable_NameInfo, proxyNamespace, proxyDomain string,
-	proxyDomainParts []string) []string {
-	out := make([]string, 0)
-	out = append(out, hostname+".")
+	proxyDomainParts []string) map[string]struct{} {
+	out := make(map[string]struct{})
+	out[hostname+"."] = struct{}{}
 	// do not generate alt hostnames if the service is in a different domain (i.e. cluster) than the proxy
 	// as we have no way to resolve conflicts on name.namespace entries across clusters of different domains
 	if proxyDomain == "" || !strings.HasSuffix(hostname, proxyDomain) {
 		return out
 	}
-	out = append(out, nameinfo.Shortname+"."+nameinfo.Namespace+".")
+	out[nameinfo.Shortname+"."+nameinfo.Namespace+"."] = struct{}{}
 	if proxyNamespace == nameinfo.Namespace {
-		out = append(out, nameinfo.Shortname+".")
+		out[nameinfo.Shortname+"."] = struct{}{}
 	}
 	// Do we need to generate entries for name.namespace.svc, name.namespace.svc.cluster, etc. ?
 	// If these are not that frequently used, then not doing so here will save some space and time
 	// as some people have very long proxy domains with multiple dots
 	// For now, we will generate just one more domain (which is usually the .svc piece).
-	out = append(out, nameinfo.Shortname+"."+nameinfo.Namespace+"."+proxyDomainParts[0]+".")
+	out[nameinfo.Shortname+"."+nameinfo.Namespace+"."+proxyDomainParts[0]+"."] = struct{}{}
 	return out
 }
 
-func (table *LookupTable) lookupHostIPv4(host string) []dns.RR {
-	ans := table.name4[host]
-	if len(ans) == 0 {
-		ans = table.cname[host]
+// Given a host, this function first decides if the host is part of our service registry.
+// If it is not part of the registry, return nil so that caller queries upstream. If it is part
+// of registry, we will look it up in one of our tables, failing which we will return NXDOMAIN.
+func (table *LookupTable) lookupHost(qtype uint16, hostname string) ([]dns.RR, bool) {
+	var hostFound bool
+	if _, hostFound = table.allHosts[hostname]; !hostFound {
+		// this is not from our registry
+		return nil, false
 	}
-	return ans
-}
 
-func (table *LookupTable) lookupHostIPv6(host string) []dns.RR {
-	ans := table.name6[host]
-	if len(ans) == 0 {
-		ans = table.cname[host]
+	var out []dns.RR
+	// Odds are, the first query will always be an expanded hostname
+	// (productpage.ns1.svc.cluster.local.ns1.svc.cluster.local)
+	// So lookup the cname table first
+	cn := table.cname[hostname]
+	if len(cn) > 0 {
+		// this was a cname match
+		hostname = cn[0].(*dns.CNAME).Target
 	}
-	return ans
+	var ipAnswers []dns.RR
+	switch qtype {
+	case dns.TypeA:
+		ipAnswers = table.name4[hostname]
+	case dns.TypeAAAA:
+		ipAnswers = table.name6[hostname]
+		// TODO: handle PTR records for reverse dns lookups
```

Comment:
if its a different type, should we return `nil, false` so we always query upstream?

Reply:
Otherwise we drop all other query types if the host is found right?

Reply:
ah true